http-server/src/main/java/io/micronaut/http/server/exceptions/response/DefaultHtmlErrorResponseBodyProvider.java

```diff
@@ -0,0 +1,190 @@
+                                    @media(min-width: 48em) {
+                                      display: inline;
+                                    }            
+                                  }          
+            """;
+
+    private final MessageSource messageSource;
+    private final LocaleResolver<HttpRequest<?>> localeResolver;
+    private final Map<LocaleStatus, String> cache = new ConcurrentHashMap<>();
+
+    DefaultHtmlErrorResponseBodyProvider(MessageSource messageSource,
+                                         LocaleResolver<HttpRequest<?>> localeResolver) {
+        this.messageSource = messageSource;
+        this.localeResolver = localeResolver;
+    }
+
+    @Override
+    public String body(@NonNull ErrorContext errorContext, @NonNull HttpResponse<?> response) {
+        final HttpStatus status = response.status();
+        Locale locale = localeResolver.resolveOrDefault(errorContext.getRequest());
+        return cache.computeIfAbsent(new LocaleStatus(locale, status), key -> html(locale, status, errorContext));
+    }
+
+    private String html(Locale locale, HttpStatus status, ErrorContext errorContext) {
+        final String errorTitleCode = status.getCode() + ".error.title";
+        final String errorTitle = messageSource.getMessage(errorTitleCode, status.getReason(), locale);
+        String header = "<h1>" + errorTitle + "</h1>";
+        header += "<h2>" + status.getCode() + "</h1>";
+        return MessageFormat.format("<!doctype html><html lang=\"en\"><head><title>{0} — {1}</title><meta charset=\"utf-8\"><meta name=\"viewport\" content=\"initial-scale=1, width=device-width\"><meta name=\"robots\" content=\"noindex, nofollow\"><style>{2}</style></head><body><main><header>{3}</header><article>{4}</article></main></body></html>",
+                status.getCode(),
+                errorTitle,
+                CSS,
+                header,
+                article(locale, status, errorContext));
+    }
+
+    private String article(Locale locale, HttpStatus status, ErrorContext errorContext) {
+        final String errorBoldCode = status.getCode() + ".error.bold";
+        final String errorCode = status.getCode() + ".error";
+        String defaultErrorBold = DEFAULT_ERROR_BOLD.get(status);
+        String defaultError = DEFAULT_ERROR.get(status);
+        String errorBold = defaultErrorBold != null ? messageSource.getMessage(errorBoldCode, defaultErrorBold, locale) : messageSource.getMessage(errorBoldCode, locale).orElse(null);
+        String error = defaultError != null ? messageSource.getMessage(errorCode, defaultError, locale) : messageSource.getMessage(errorCode, locale).orElse(null);
+        StringBuilder sb = new StringBuilder();
+
+        for (io.micronaut.http.server.exceptions.response.Error e : errorContext.getErrors()) {
+            if (!e.getMessage().equalsIgnoreCase(status.getReason())) {
+                sb.append(e.getMessage());
```

Comment:
XSS

Reply:
not sure how Cross Site Scripting (XSS) attack could be done because we are outputting the `Error::getMessage` in the error processor. Are we afraid of leaking internals of the application to a potential attacker? 

Please, note [we are already doing this](https://github.com/micronaut-projects/micronaut-core/blob/4.7.x/http-server/src/main/java/io/micronaut/http/server/exceptions/response/HateoasErrorResponseProcessor.java#L59) for JSON responses.

Reply:
The error message can potentially contain user-controlled data which may be HTML which can be used for XSS

Reply:
so, you are suggesting that we don't output non standard reasons ?

Reply:
Either that, or sanitize/escape them

Reply:
>  user-controlled data 

is not the programmer of the application the one who populates custom status reasons?

Reply:
Yep but we can't rely on them to only do sensible things

Reply:
I don't see the difference between a programmer outputting a custom status reason or a programmer outputting a custom response in a controller. He could create a security vulnerability in any case. I don't see how a custom status reason can be leverage as an attack vector from an outside user.

Reply:
We could introduce [an API to sanitize HTML ](https://github.com/micronaut-projects/micronaut-security/issues/1817).

Reply:
To avoid [injection vulnerabilities](https://owasp.org/www-community/Injection_Flaws) (not just XSS), when outputting a structured format, strings must be validated, sanitized or escaped, unless they are from a limited set of strings known not be harmful (e.g. the reason phrases of the HttpStatus enum). This principle applies everywhere in the framework, such as HTTP/1 header generation, JSON generation, form generation etc. 

Where the strings come from is secondary. We cannot take strings supplied by the app programmer and output them unescaped.

Reply:
fwiw escaping should be very easy, especially since this is html 5: https://stackoverflow.com/questions/7381974/which-characters-need-to-be-escaped-in-html

Reply:
> We cannot take strings supplied by the app programmer and output them unescaped.

If a malicious programmer returns a malicious HTML in his controller method, we don't escape that. 

I just don't see how someone could leverage custom status as an attack unless they have complete control of writing application code. At that point, I don't think we can do anything. If he wants to write a malicious output, we could do that in a million ways.  

I will try to create an API for HTML sanitisation. But I don't see how to leverage custom status code as an attack.

Reply:
JSON analogy: If a programmer has a controller like `@Get String foo() { return "\"foo\""; }`, then no, we don't escape the JSON. If they define a controller that returns a structure *containing* a string though, like `@Get MyBean foo() { return new MyBean("\"foo\""); }`, then the JSON contains the string escaped, i.e. `{"field":"\"foo\""}`.

It's a basic principle used to avoid injection vulnerabilities. "Error" is a complex structure containing a string, not a string that is explicitly meant as raw HTML, so it must be escaped.

Reply:
I have sanitized error and status https://github.com/micronaut-projects/micronaut-core/pull/11210/commits/3f95b0cb0d3dababfa22c13606450363037f3fc4